src/centralized_telescope.rs

```diff
@@ -0,0 +1,419 @@
+    /// - H2(t, x_0, ..., x_u) = true
+    fn dfs(
+        setup: &Setup,
+        bins: &[Vec<Element>],
+        round: &Round,
+        limit: u64,
+    ) -> (u64, Option<Proof>) {
+        if round.s_list.len() as u64 == setup.u {
+            if Proof::h2(setup, round) {
+                let v = round.v;
+                let t = round.t;
+                let items = round.s_list.clone();
+                return (limit, Some(Proof { v, t, items }));
+            } else {
+                return (limit, None);
+            }
+        }
+
+        bins[round.h_u64 as usize]
+            .iter()
+            .fold((limit, None), |(l, proof_opt), &s| {
+                if proof_opt.is_some() || l == setup.b {
+                    (l, proof_opt)
+                } else {
+                    Self::dfs(setup, bins, &Round::update(round, s), l + 1)
+                }
+            })
+    }
+
+    /// Indexed proving algorithm, returns an empty proof if no suitable
+    /// candidate is found within the setup.b steps.
+    fn prove_index(setup: &Setup, set: &[Element], v: u64) -> (u64, Option<Proof>) {
+        let mut bins: Vec<Vec<Element>> = vec![vec![]; setup.n_p as usize];
+        for &s in set.iter() {
+            bins[Proof::h0(setup, v, s) as usize].push(s);
+        }
+
+        (0..setup.d).fold((0, None), |(limit, proof_opt), t| {
+            if proof_opt.is_some() || limit == setup.b {
+                (limit, proof_opt)
+            } else {
+                let round = Round::new(v, t, setup.n_p);
+                Proof::dfs(setup, &bins, &round, limit + 1)
+            }
+        })
+    }
+
+    /// Alba's proving algorithm, based on a depth-first search algorithm.
+    /// Calls up to setup.r times the prove_index function and returns an empty
+    /// proof if no suitable candidate is found.
+    pub fn prove(setup: &Setup, set: &[Element]) -> Option<Self> {
+        (0..setup.r).find_map(|v| Proof::prove_index(setup, set, v).1)
+    }
+
+    /// Alba's proving algorithm used for benchmarking, returning a proof as
+    /// well as the number of  steps ran to find it.
+    pub fn bench(setup: &Setup, set: &[Element]) -> (u64, u64, Option<Self>) {
+        (0..setup.r).fold((0, setup.r, None), |(limit, r, proof_opt), v| {
+            if proof_opt.is_some() {
+                (limit, r, proof_opt)
+            } else {
+                let (l, opt) = Proof::prove_index(setup, set, v + 1);
+                (limit + l, r, opt)
+            }
+        })
+    }
+
+    /// Alba's verification algorithm, follows proving algorithm by running the
+    /// same depth-first search algorithm.
+    pub fn verify(setup: &Setup, proof: Proof) -> bool {
+        if proof.t >= setup.d || proof.v >= setup.r || proof.items.len() as u64 != setup.u {
+            return false;
+        }
+        let r0 = Round::new(proof.v, proof.t, setup.n_p);
+        let (b, round) = proof.items.iter().fold((true, r0), |(b, r), &s| {
```

Comment:
idem - instead of `fold`, you can use a `for` loop to both improve readability of the code but also allow for early termination (instead of keeping a boolean in the accumulator). Simply return `false` at the iteration when you find a `h0` mismatch and return early.